### DISCUSSION TITLE
MAME subsystems

### DISCUSSION THREAD

**Discussion by brunobelo** [OnionUI/Onion#485] [Ideas]
Ok, first of all, kudos to all involved with this project. I recently got my hands on a Miyoo Mini and I'm in love with it. And with onion, it's even better!

I already made a theme for myself, with the logos of the consoles (which means copyrighted material) and no system name written below them, it's awesome, if somebody wants it I can send. But it's not a plug and play theme (you'll have to edit the emus in order to make it work), so be aware. 

My question is: I want to run some MAME subsystems on it, and I know it's possible since I've done it on every single platform I tried, including raspberry, Batocera, Android, Linux (x86 and x64) or Windows. I know what I have to do, being:

*The BIOS files are in place, mostly on the same folder as the ROMs;
*The subsystem name is correct since they are the same as I use on my cellphone's RetroArch;
*The hashes are at the right place (I guess): sdcard/BIOS/mame2015/hash (on the phone they are at /RetroArch/system/mame2015/hash);
*I'm running the mame2015 core and it works for regular MAME roms;
*The retroarch's log would not give me any error, just crash;
*I was unable to run mame (current) on it, I don't know why. 

I want to run AdventureVision, Super Cassette Vision or even Casio PV-1000 on this little thing, and I guess it's possible. :)

Can someone help me on that?

**Comment by dwmccoy** [comment]
Okay,

I have tried to make Adventure Vision work.  Here are my findings.
From what I can find, Adventure Vision requires the following BIOS files
b225__ins8048-11kdp_n.u5
b8223__cop411l-kcn_n.u8
From what I can find they are supposed be placed into a zip file name advision.zip

I installed mame2015 into /sdcard/RApp/mame2015
My config.json file looks for ROMS in /sdcard/RRoms/mame2015

Inside of the sdcard/RRoms/mame2015 folder I made a folder called advision
I placed my Adventure Vision Game Roms into the /sdcard/RRoms/mame2015/advision folder
I have place my BIOS file named advision.zip into the following directories in an attempt to make it work 
/sdcard/BIOS/mame2015
/sdcard/BIOS/mame2015/advision
/sdcard/RRoms/mame2015/advision
/sdcard/RRoms/mame2015/advision/advision

Whenever I launch a Game Rom, this is the contents of the Retroarch log file

[INFO] [Config]: Loading default config.
[INFO] [Config]: Looking for config in: "/mnt/SDCARD/RetroArch//.retroarch/retroarch.cfg".
[INFO] === Build =======================================
[INFO] CPU Model Name: ARMv7 Processor rev 5 (v7l)

[INFO] Capabilities:  NEON VFPv3 VFPv4
[INFO] Built: May  3 2022
[INFO] Version: 1.10.3
[INFO] =================================================
[INFO] [Input]: Found input driver: "sdl_dingux".
[INFO] [Core]: Loading dynamic libretro core from: "/mnt/SDCARD/RApp/mame2015/mame2015_libretro.so"
[INFO] [Overrides]: No core-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/MAME 2015.cfg".
[INFO] [Overrides]: No content-dir-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/advision.cfg".
[INFO] [Overrides]: No game-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/Space Force (USA, Europe).cfg".
[INFO] [Environ]: SET_VARIABLES.
[INFO] [Remaps]: Remap directory: "/mnt/SDCARD/Saves/CurrentProfile/config/remaps".
[INFO] [Overrides]: Redirecting save file to "/mnt/SDCARD/Saves/CurrentProfile/saves/MAME 2015/Space Force (USA, Europe).srm".
[INFO] [Overrides]: Redirecting save state to "/mnt/SDCARD/Saves/CurrentProfile/states/MAME 2015/Space Force (USA, Europe).state".
[INFO] [Environ]: GET_LOG_INTERFACE.
[INFO] [Environ]: SYSTEM_DIRECTORY: "/mnt/SDCARD/BIOS".
[libretro INFO] SYSTEM_DIRECTORY: /mnt/SDCARD/BIOS[INFO] [Environ]: CORE_ASSETS_DIRECTORY: "/mnt/SDCARD/RetroArch/.retroarch/downloads".
[libretro INFO] CONTENT_DIRECTORY: /mnt/SDCARD/RetroArch/.retroarch/downloads[INFO] [Environ]: GET_SAVE_DIRECTORY.
[libretro INFO] SAVE_DIRECTORY: /mnt/SDCARD/Saves/CurrentProfile/saves/MAME 2015[INFO] [Content]: Content loading skipped. Implementation will load it on its own.
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] [State]: found_last_state_slot: #0
[INFO] [SRAM]: Skipping SRAM load.
[INFO] [Core]: Version of libretro API: 1, Compiled against API: 1
[libretro INFO] AV_INFO: width=320 height=240
[libretro INFO] AV_INFO: max_width=1600 max_height=1200
[libretro INFO] AV_INFO: aspect_ratio = 0.000000
[libretro INFO] AV_INFO: fps = 60.000000 sample_rate = 48000.000000
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Set video size to: 320x240.
[INFO] [Joypad]: Found joypad driver: "sdl_dingux".
[INFO] [Video]: Found display server: "null".
[INFO] [OSS]: Using audioserver.
[INFO] [Display]: Found display driver: "null".
[INFO] [Playlist]: Loading history file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_history.lpl".
[INFO] [Playlist]: Loading history file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_music_history.lpl".
[INFO] [Playlist]: Loading favorites file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_favorites.lpl".
[INFO] [Playlist]: Written to playlist file: /mnt/SDCARD/Saves/CurrentProfile/lists/content_history.lpl
[libretro INFO] Starting game:/mnt/SDCARD/RApp/mame2015/../../RRoms/mame2015/advision/Space Force (USA, Europe).zip
[libretro WARN] Driver Space Force (USA, Europe) not found -1
[libretro ERROR] Driver not found: Space Force (USA, Europe)


Every game rom errors in the same way.  This is also the same behavior of the included mame2003-plus core.  Maybe someone else has a better idea.

**Reply by brunobelo** [reply]
I'll look further into it, thanks a lot for your response!

**Reply by brunobelo** [reply]
My log does not have this error... I'll upload it here asap. Interesting, your error is like someone removed the driver for the subsystem you're trying to run... I saw batocera dudes done this before, I had to recompile the mame core to use the subsystems because they remove this function inexplicably...

**Reply by brunobelo** [reply]
Looking into it again, looks like the rom name is wrong, it should be spacefrc.zip, can you try to rename it and see if it runs? Also do you have the hashes? I never could run a subsystem without hashes...

**Reply by dwmccoy** [reply]
downloaded the advision.xml hash from https://github.com/mamedev/mame/tree/master/hash
didn't know where to place the file so I placed it into the following directories hoping one would be a winner.

sdcard/BIOS/mame2015/hash/advision.xml
sdcard/BIOS/mame2015/advision/hash/advision.xml
sdcard/RetroArch/.retroarch/system/mame2015/hash/advision.xml
sdcard/RRoms/mame2015/hash/advision.xml
sdcard/RRoms/mame2015/advision/hash/advision.xml
sdcard/RRoms/mame2015/advision/advision/hash/advision.xml

The error is the same

[libretro INFO] Starting game:/mnt/SDCARD/RApp/mame2015/../../RRoms/mame2015/advision/spacefrc.zip
[libretro WARN] Driver spacefrc not found -1
[libretro ERROR] Driver not found: spacefrc

Also note that when I added Adventure Vision defender.zip to the sdcard/RRoms/mame2015/advision folder, the Mame2015 core seen the game as Defender (Red label).  That would lead me to believe it thinks I added the Arcade version of Defender and not the Adventure Vision version.  It crashes back to the rom list and generates this output instead.

[INFO] [Input]: Found input driver: "sdl_dingux".
[INFO] [Core]: Loading dynamic libretro core from: "/mnt/SDCARD/RApp/mame2015/mame2015_libretro.so"
[INFO] [Overrides]: No core-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/MAME 2015.cfg".
[INFO] [Overrides]: No content-dir-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/advision.cfg".
[INFO] [Overrides]: No game-specific overrides found at "/mnt/SDCARD/Saves/CurrentProfile/config/MAME 2015/defender.cfg".
[INFO] [Environ]: SET_VARIABLES.
[INFO] [Remaps]: Remap directory: "/mnt/SDCARD/Saves/CurrentProfile/config/remaps".
[INFO] [Overrides]: Redirecting save file to "/mnt/SDCARD/Saves/CurrentProfile/saves/MAME 2015/defender.srm".
[INFO] [Overrides]: Redirecting save state to "/mnt/SDCARD/Saves/CurrentProfile/states/MAME 2015/defender.state".
[INFO] [Environ]: GET_LOG_INTERFACE.
[INFO] [Environ]: SYSTEM_DIRECTORY: "/mnt/SDCARD/BIOS".
[libretro INFO] SYSTEM_DIRECTORY: /mnt/SDCARD/BIOS[INFO] [Environ]: CORE_ASSETS_DIRECTORY: "/mnt/SDCARD/RetroArch/.retroarch/downloads".
[libretro INFO] CONTENT_DIRECTORY: /mnt/SDCARD/RetroArch/.retroarch/downloads[INFO] [Environ]: GET_SAVE_DIRECTORY.
[libretro INFO] SAVE_DIRECTORY: /mnt/SDCARD/Saves/CurrentProfile/saves/MAME 2015[INFO] [Content]: Content loading skipped. Implementation will load it on its own.
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] [State]: found_last_state_slot: #0
[INFO] [SRAM]: Skipping SRAM load.
[INFO] [Core]: Version of libretro API: 1, Compiled against API: 1
[libretro INFO] AV_INFO: width=320 height=240
[libretro INFO] AV_INFO: max_width=1600 max_height=1200
[libretro INFO] AV_INFO: aspect_ratio = 0.000000
[libretro INFO] AV_INFO: fps = 60.000000 sample_rate = 48000.000000
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Set video size to: 320x240.
[INFO] [Joypad]: Found joypad driver: "sdl_dingux".
[INFO] [Video]: Found display server: "null".
[INFO] [OSS]: Using audioserver.
[INFO] [Display]: Found display driver: "null".
[INFO] [Playlist]: Loading history file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_history.lpl".
[INFO] [Playlist]: Loading history file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_music_history.lpl".
[INFO] [Playlist]: Loading favorites file: "/mnt/SDCARD/Saves/CurrentProfile/lists/content_favorites.lpl".
[INFO] [Playlist]: Written to playlist file: /mnt/SDCARD/Saves/CurrentProfile/lists/content_history.lpl
[libretro INFO] Starting game:/mnt/SDCARD/RApp/mame2015/../../RRoms/mame2015/advision/defender.zip
[libretro INFO] Game name: defender, Game description: Defender (Red label)
[libretro INFO] Creating frontend for game: defender
[libretro INFO] Softlists: 0
[libretro INFO] Screen orientation: HORIZONTAL
[libretro INFO] OSD initialization complete


Hope this helps you out a bit more.

**Reply by brunobelo** [reply]
It's helping. :)

The correct hash folder must be "BIOS/mame2015/hash/*.xml", you used a correct one. 

Maybe you're right, mame is "thinking" that Defender is the arcade version, but that's why the folder name and hash is for... And I don't see an error on your log, just like mine. :-\

**Comment by Olywa** [comment]
I'm not sure of the answers, it sounds like you are doing everything right. 

The only thing I can advise is that I know that Mame Current will not work due to the hardware limitations of the Mini (the Mame Current core is I think ~300mb before a game is even loaded and the Mini has only 128mb RAM). Maybe there are similar issues with the 2015 core (not sure of the size) but then that wouldn't explain how you managed to get Mame 2015 arcade roms to load! :o/

**Reply by brunobelo** [reply]
It's an interesting thought, the amount of RAM. Maybe you're right...
